pkg/api/auth/auth_handler.go

```diff
@@ -0,0 +1,87 @@
+package auth
+
+import (
+	"net/http"
+
+	"github.com/rancher/harvester/pkg/auth/jwe"
+	"github.com/rancher/harvester/pkg/auth/jwt"
+	"github.com/rancher/harvester/pkg/config"
+
+	"github.com/pkg/errors"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/client-go/tools/clientcmd/api"
+)
+
+const (
+	jwtServiceAccountNamespaceKey = "kubernetes.io/serviceaccount/namespace"
+	jwtServiceAccountNameKey      = "kubernetes.io/serviceaccount/service-account.name"
+	jwtServiceAccountUIDKey       = "kubernetes.io/serviceaccount/service-account.uid"
+)
+
+func NewMiddleware(scaled *config.Scaled) (*Middleware, error) {
+	keyHolder, err := jwe.NewRSAKeyHolderWithSecret(scaled.CoreFactory.Core().V1().Secret())
+	if err != nil {
+		return nil, err
+	}
+
+	tokenManager := jwe.NewTokenManager(keyHolder)
+	return &Middleware{
+		tokenManager: tokenManager,
+	}, nil
+}
+
+type Middleware struct {
+	tokenManager *jwe.TokenManager
+}
+
+func (h *Middleware) AuthMiddleware(rw http.ResponseWriter, r *http.Request, handler http.Handler) {
+	jweToken, err := extractJWETokenFromRequest(r)
+	if err != nil {
+		rw.WriteHeader(http.StatusUnauthorized)
+		rw.Write(responseBody(TokenResponse{Errors: []string{err.Error()}}))
+		return
+	}
+
+	authInfo, err := h.tokenManager.Decrypt(jweToken)
+	if err != nil {
+		rw.WriteHeader(http.StatusUnauthorized)
+		rw.Write(responseBody(TokenResponse{Errors: []string{err.Error()}}))
+		return
+	}
+
+	userInfo, err := getUserFromAuthInfo(authInfo)
+	if err != nil {
+		rw.WriteHeader(http.StatusUnauthorized)
+		rw.Write(responseBody(TokenResponse{Errors: []string{err.Error()}}))
+		return
+	}
+
+	ctx := request.WithUser(r.Context(), userInfo)
+	r = r.WithContext(ctx)
+	handler.ServeHTTP(rw, r)
+}
+
+func getUserFromAuthInfo(authInfo *api.AuthInfo) (user.Info, error) {
+	if authInfo.Token != "" {
+		claims, err := jwt.GetJWTTokenClaims(authInfo.Token)
+		if err != nil {
+			return nil, err
+		}
+		name, ok1 := claims[jwtServiceAccountNameKey]
+		namespace, ok2 := claims[jwtServiceAccountNamespaceKey]
+		uid, ok3 := claims[jwtServiceAccountUIDKey]
+		if ok1 && ok2 && ok3 {
+			return &user.DefaultInfo{
+				Name: serviceAccountName(namespace.(string), name.(string)),
+				UID:  uid.(string),
+			}, nil
+		}
+	}
+
+	if len(authInfo.ClientCertificateData) != 0 && len(authInfo.ClientKeyData) != 0 {
+		return adminUser(), nil
+	}
```

Comment:
set login user as an admin by only validating its certificate data and key not empty is still doubtable.

Reply:
could decode userName and group info from the certificate.